controller/cache/cache.go

```diff
@@ -908,3 +959,218 @@ func (c *liveStateCache) GetClusterCache(server *appv1.Cluster) (clustercache.Cl
 func (c *liveStateCache) UpdateShard(shard int) bool {
 	return c.clusterSharding.UpdateShard(shard)
 }
+
+// MarkClusterTainted marks a cluster as having tainted cache state
+func (c *liveStateCache) MarkClusterTainted(server string, reason string, gvk string, errorType string) {
+	c.taintManager.markTainted(server, gvk, errorType, reason)
+}
+
+// IsClusterTainted checks if a cluster is in tainted state
+func (c *liveStateCache) IsClusterTainted(server string) bool {
+	return c.taintManager.isTainted(server)
+}
+
+// GetTaintedGVKs returns a list of tainted GVKs for a cluster
+func (c *liveStateCache) GetTaintedGVKs(server string) []string {
+	return c.taintManager.getTaintedGVKs(server)
+}
+
+// ClearClusterTaints removes all taints for a cluster
+func (c *liveStateCache) ClearClusterTaints(server string) {
+	c.taintManager.clearTaints(server)
+}
+
+// shouldReturnPartialCache determines whether to return partial cache based on
+// tainted GVKs and error type. Only returns partial cache when:
+// 1. Specific GVKs are tainted (indicating partial, not total failure)
+// 2. The error is a recoverable cache error (not a connection failure)
+func (c *liveStateCache) shouldReturnPartialCache(server string, err error) bool {
+	taintedGVKs := c.taintManager.getTaintedGVKs(server)
+
+	// Only return partial cache if we have specific GVK failures AND it's a recoverable error
+	if len(taintedGVKs) == 0 || !isPartialCacheError(err) {
+		return false
+	}
+
+	log.WithField("cluster", server).WithField("taintedGVKs", taintedGVKs).Debug("Partial cache acceptable for tainted GVKs")
+	return true
+}
+
+// Pre-compiled regex patterns for better performance
+var (
+	partialCacheErrorPatterns = []*regexp.Regexp{
+		// Conversion webhook errors - specific to certain GVKs
+		regexp.MustCompile(`(?i)conversion\s+webhook.*(?:failed|error)`),
+
+		// Pagination token expiration - recoverable
+		regexp.MustCompile(`(?i)expired.*too\s+old\s+resource\s+version`),
+
+		// Network timeout errors - often transient and recoverable
+		regexp.MustCompile(`(?i)connection\s+reset\s+by\s+peer`),
+		regexp.MustCompile(`(?i)i/o\s+timeout`),
+
+		// TODO: Consider adding resource list failures after reproducing the issue
+		// Maintainer noted this as a potential partial failure case, but we don't have
+		// a working reproduction to validate the behavior yet.
```

Comment:
@crenshaw-dev mentioned that when the auth token times out on a very long list operation, we might enter this degraded state.  We'd want to catch that here once we know what that error looks like and can reproduce it.